keras/layers/attention.py

```diff
@@ -0,0 +1,609 @@
+
+        return output, new_cell_states, new_attention_states
+
+    @staticmethod
+    def _num_elements(x):
+        if hasattr(x, '__len__'):
+            return len(x)
+        else:
+            return 1
+
+    @property
+    def _num_wrapped_states(self):
+        return self._num_elements(self.cell.state_size)
+
+    @property
+    def _num_attention_states(self):
+        return self._num_elements(self.attention_state_size)
+
+    def build(self, input_shape):
+        """Builds attention mechanism and wrapped cell (if keras layer).
+
+        Arguments:
+            input_shape: list of tuples of integers, the input feature shape
+                (inputs sequence shape without time dimension) followed by
+                constants (i.e. attended) shapes.
+        """
+        if not isinstance(input_shape, list):
+            raise ValueError('input shape should contain shape of both cell '
+                             'inputs and constants (attended)')
+
+        attended_shape = input_shape[1:]
+        input_shape = input_shape[0]
+        self.attended_spec = [InputSpec(shape=shape) for shape in attended_shape]
+        if isinstance(self.cell.state_size, int):
+            cell_state_size = [self.cell.state_size]
+        else:
+            cell_state_size = list(self.cell.state_size)
+        self.attention_build(
+            input_shape=input_shape,
+            cell_state_size=cell_state_size,
+            attended_shape=attended_shape,
+        )
+
+        if isinstance(self.cell, Layer):
+            cell_input_shape = (input_shape[0],
+                                self.attention_size +
+                                input_shape[-1] if self.concatenate_input
+                                else self._attention_size)
+            self.cell.build(cell_input_shape)
+
+        self.built = True
+
+    def compute_output_shape(self, input_shape):
+        if hasattr(self.cell.state_size, '__len__'):
+            cell_output_dim = self.cell.state_size[0]
+        else:
+            cell_output_dim = self.cell.state_size
+
+        return input_shape[0], cell_output_dim
+
+    @property
+    def trainable_weights(self):
+        return super(_RNNAttentionCell, self).trainable_weights + \
+               self.cell.trainable_weights
+
+    @property
+    def non_trainable_weights(self):
+        return super(_RNNAttentionCell, self).non_trainable_weights + \
+               self.cell.non_trainable_weights
+
+    def get_config(self):
+        config = {'attend_after': self.attend_after,
+                  'concatenate_input': self.concatenate_input}
+
+        cell_config = self.cell.get_config()
+        config['cell'] = {'class_name': self.cell.__class__.__name__,
+                          'config': cell_config}
+        base_config = super(_RNNAttentionCell, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+class MixtureOfGaussian1DAttention(_RNNAttentionCell):
```

Comment:
Need unit tests for `MixtureOfGaussian1DAttention`

Reply:
absolutely, was waiting for feedback on the overall approach.